### PR TITLE
implement RFCs/294 ; disallow enum <=> enum conversion

### LINKED ISSUE #294: disallow enum to enum conversion (require passing through `ord`)
## proposal
```nim
when true:
  type Koo = enum k1, k2
  type Goo = enum g1, g2
  discard k2.Goo # make this illegal (currently legal)
  discard k2.ord.Goo # this stays legal
```

rationale: this seems error prone and it's more explicit to pass through `ord` (and should result in same C code)

### PR DESCRIPTION
**Pull request by timotheecour**
* fix https://github.com/nim-lang/RFCs/issues/294
* allow `-d:nimLegacyConvEnumEnum`

## CI failures caused by this:
only those 2:
* nimgame2
/Users/runner/work/Nim/Nim/pkgstemp/nimgame2/nimgame2/texturegraphic.nim(241, 41) Error: type mismatch: got <Flip> but expected 'RendererFlip = enum'
* nimquery
/Users/runner/work/Nim/Nim/pkgstemp/nimquery/nimquery.nim(915, 68) Error: type mismatch: got <TokenKind> but expected 'Combinator = enum'

these packages can be fixed upstream, or testament can add logic to pass flags (eg `-d:nimLegacyConvEnumEnum` or everytime such kind of issue shows up) to tests that need custom flags; this can be done by overriding `XDG_CONFIG_HOME` for those tests, pointing to a custom config.nims, so that these flags apply recursively through nim invocations via `exec`. (EDIT: see https://github.com/nim-lang/RFCs/issues/336)

## future work
- [x] reenable nimquery in important_packages after https://github.com/GULPF/nimquery/pull/10 gets merged => https://github.com/nim-lang/Nim/pull/17936

### CODE REVIEW

**Comment by cooldome** [comment]
`cast` from one enum to another should work, please test. IMO, cast should work even if enums have different byte size.

Should changelog recommend `cast` instead of `ord`?

**Comment by timotheecour** [comment]
> cast from one enum to another should work, please test. IMO, cast should work even if enums have different byte size.

done

> Should changelog recommend cast instead of ord?

no, cast is overkill in this use case, and won't help catch bugs eg:
`cast[Foo](bar)` would happily cast any type to any type. `ord` should have 0 overhead.

File: changelog.md [hunk 1]
```diff
@@ -69,6 +69,15 @@
...
 
+- An enum now can't be converted to another enum directly, you must use `ord` (or `cast`, but
+  compiler won't help if you misuse it).
+  ```
+  type A = enum a1, a2
+  type B = enum b1, b2
+  doAssert not compiles(a1.B)
+  doAssert compiles(a1.ord.B)
+  ```
+  for a transition period, use `-d:nimLegacyConvEnumEnum`.
```
Comment (dom96) [inline changelog.md]: I think this should be put under a "Breaking Changes" heading

Comment (timotheecour) [reply changelog.md]: i agree but there used to be one in changelog template and it got removed

Comment (timotheecour) [reply changelog.md]: done, added: `Changes affecting backward compatibility`

**Comment by timotheecour** [comment]
PTAL

File: changelog.md [hunk 2]
```diff
@@ -2,6 +2,38 @@
 
 
 
+## Changes affecting backward compatibility
```
Comment (timotheecour) [inline changelog.md:5]: that section was requested by @dom96 in another PR IIRC

**Comment by timotheecour** [comment]
@Araq PTAL (fixed bitrot)

File: changelog.md (hunk 2 above)
Comment (timotheecour) [reply changelog.md:5]: EDIT: actually it was this very PR see https://github.com/nim-lang/Nim/pull/16351#discussion_r545386949

**Comment by timotheecour** [comment]
ping @Araq I had to rebase again to fix bitrot

**Comment by alaviss** [comment]
Can we have a migration error message for this? This will be extremely confusing for anyone not tracking devel and not having their code in `important_packages` (I have to debug one such codebase recently).

I'm a bit worried how we keep on introducing tiny breaking changes in the language spec between versions.

**Comment by timotheecour** [comment]
that's what `-d:nimLegacyConvEnumEnum` + the changelog section `Changes affecting backward compatibility` is for, those changes are made for a reason. 

The changelog clearly mentions this change. In any case, I've just sent out https://github.com/nim-lang/Nim/pull/17928 for this case, but in other cases, it may not always be practical to do this (see that section in the changelog). Users who upgrade should read that section.

**Comment by Clyybber** [comment]
I'm not sure anymore wether this change is a good idea, the user is already explicit when writing `foo.Bar`, having to add `.ord` in there is unneccessary syntax salt.
I was originally positive about this change because two enums might not have any meaningful relation and this would make conversions between them more explicit, but thinking more about it I don't think it fits well with the rest of the language; conversions from `distinct T` to `T` or vice versa don't have this extra explicitness either, even though the value of a `distinct T` might not have any meaningful relation to it's value as `T`.

**Comment by timotheecour** [comment]
(let's followup in https://github.com/nim-lang/RFCs/issues/294#issuecomment-831472803 instead to avoid duplicating the discussion)

**Pull request merged**